keps/sig-scheduling/1923-try-nominated-node-first/README.md

```diff
@@ -0,0 +1,156 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+If the scheduler fails to fit an incoming pod on any node, the scheduler will try to preempt lower
+priority pods running on a selected node to make room for the pod. The name of this node will be set
+in the pod's `pod.Status.NominatedNodeName`.
+
+The Node is called *Nominated* to indicate the intent for the Pod to be scheduled on it once preemption
+of other Pods finish. However, the `Pod.status.nominatedNodeName` information is not directly used in
+the Pod's following scheduling attempts.
+
+Pod scheduling is split into two phases, the scheduling cycle and the binding cycle, the scheduling cycle
+primarily includes filtering and scoring.
+
+When preemption happens in a previous scheduling cycle, there is a high chance that the nominated node is
+the *only* node that satisfies the filters for the unscheduled Pod that triggered preemption.
+
+This KEP proposes to change the scheduling cycle such that nominated node of a pod is evaluated first
+and schedule the pod on that node if it fits. If the nominated node doesn't fit the pod, only then the
+scheduling cycle continues with the standard logic of evaluating the rest of the nodes in the cluster.
+
+## Motivation
+
+In real production environment, pods can have different priorites due to business needs, the preemption
+could happen to make sure higher priority pods could get scheduled.
+
+In cluster with large number of computing nodes, evaluating all nodes when scheduling a pod is time consuming.
+
+### Goals
+
+In the case where `pod.Status.NominatedNodeName` is set for an incoming pod, the scheduler will evaluate the
+nominated node first; if the nominated node doesn't fit the pod, the scheduling cycle will continue to evaluate
+the rest of the nodes in the cluster just like we do today.
+
+
+## Proposal
+
+### User Stories (Optional)
+
+Users want faster scheduling. Since it is highly likely the pod will only fit on the nominated node, the improvement
+in scheduling latency will come at negligible cost (the cost being placing the pod on a less optimal node).
+
+### Notes/Constraints/Caveats (Optional)
+
+When this feature is enabled the preemptor Pod might not be dispatched to the best candidated node in some corner case,
+e.g. another node releases the resources and becomes the best candidate while the victim pods got removed from the
+nominated node.
+
+## Design Details
+
+### Implementation Details
+
+1. In filtering phase, which is currently implemented in the method of `findNodesThatFitPod`, check the nominated node
+   first if the incoming pod has the `pod.Status.NominatedNodeName` defined and the feature gate is enabled.
+
+2. In case the nominated node doesn't suit for the incoming pod anymore, get `err` from `findNodesThatPassFilters` where
```

Comment:
Looks reasonable now.

However, I have one suggestion. Rather than evaluating the nominated first and only the rest of the nodes if it doesn't fit, we could do those 2 steps concurrently. Otherwise we would be increasing latency by potentially 2x in the worst case scenario.

Reply:
why 2x? it is a single node evaluated through filters only.

Reply:
If you have the number of Nodes equal to # of threads, it is 2x. But it's lower in general.

Reply:
not worth the complexity in my opinion. it is not expensive to evaluate a single node through filters.

Reply:
We still have to run PreFilter for all nodes initially, right?
In that case, yes, it might not be worth the complexity.

Reply:
Right, I assumed prefilter would be run once.

Reply:
yes, prefilter will run only once, and the times to call those filters are not changed at all.